src/v/kafka/server/handlers/fetch.cc

```diff
@@ -672,6 +673,492 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+                  });
+            }
+        }
+    }
+
+private:
+    void start_worker_aborts(const fetch_plan& plan) {
+        for (const auto& fetch : plan.fetches_per_shard) {
+            if (!fetch.empty()) {
+                _worker_aborts[fetch.shard];
+            }
+        }
+    }
+
+    ss::future<> abort_workers() {
+        return seastar::parallel_for_each(_worker_aborts, [](auto& wa) {
+            auto& [shard, as] = wa;
+            return ss::smp::submit_to(shard, [&as] { as.request_abort(); });
+        });
+    }
+
+    /**
+     * Determines if the should_stop_fetch() condition should be checked again.
+     * The return future is set if;
+     * - octx.deadline has been reached.
+     * - _as has been aborted.
+     * - one of the shard workers has returned results.
+     */
+    ss::future<> progress_conditions(op_context& octx) {
+        ss::promise<> timeout_pr, abort_pr;
+
+        // A connection can close and stop the sharded abort source before we
+        // can subscribe to it. So we check here if that is the case and return
+        // if so.
+        if (!octx.rctx.abort_source().local_is_initialized()) {
+            co_return;
+        }
+
+        auto abort_sub_opt = octx.rctx.abort_source().subscribe(
+          [&abort_pr]() noexcept { abort_pr.set_value(); });
+
+        if (!abort_sub_opt) {
+            abort_pr.set_value();
+        }
+
+        ss::timer<model::timeout_clock> s{
+          [&timeout_pr] { timeout_pr.set_value(); }};
+
+        if (octx.deadline) {
+            s.arm(octx.deadline.value());
+        }
+
+        // Ignoring any exceptions from these futures is fine since they are
+        // only used to signal the coordinator that it should check if the fetch
+        // request should end or not.
+        co_await ss::when_any(
+          ignore_exceptions(_has_completed_shard_fetches.wait()),
+          ignore_exceptions(timeout_pr.get_future()),
+          ignore_exceptions(abort_pr.get_future()));
+    }
+
+    /*
+     * `start_shard_fetch_worker` executes on the coordinator shard. It builds
+     * the `shard_local_fetch_context` struct needed to start the
+     * shard_fetch_worker then makes a cross shard call to the shard fetch
+     * worker if needed.
+     *
+     * It then waits until the shard_fetch_worker completes
+     * and then notifies the fetch coordinator.
+     */
+    ss::future<> start_shard_fetch_worker(
+      op_context& octx, shard_fetch fetch, size_t min_fetch_bytes) {
+        // if over budget skip the fetch.
+        if (octx.bytes_left <= 0) {
+            co_return;
+        }
+        // no requests for this shard, do nothing
+        if (fetch.empty()) {
+            co_return;
+        }
+
+        const bool foreign_read = fetch.shard != ss::this_shard_id();
+
+        fetch_worker::worker_result results
+          = co_await octx.rctx.partition_manager().invoke_on(
+            fetch.shard,
+            octx.ssg,
```

Comment:
Having the cross-shard request block while consuming ssg units is a big no-no I think. It means that if there are N fetch requests waiting for data from a shard (a "no work" operation) any other requests which want to flow through this ssg will be blocked. It could also lead to deadlock (though this is harder to determine).

I think we can just remove the ssg for this use (is it the only use?).

Reply:
The only use is once here and once in the old fetch path.

My understanding is that `ssg` is used to limit the number of on-going fetch tasks on a given shard. However, at this point we already have semaphores earlier in the fetch path that are used for a similar reason. So `ssg` only seems to complicate reasoning about where backpressure is coming from. Is there anything we're actually gaining by keeping it in the current polling fetch path?

Reply:
Definitely agree it should be removed from the new non-polling fetch impl. Just wondering what its actual use(and/or benefit) is in the current polling fetch impl.

Reply:
Sounds good on removal.